src/main/java/org/jenkinsci/plugins/plaincredentials/impl/FileCredentialsImpl.java

```diff
@@ -74,34 +119,124 @@
             );
         }
         if (LOGGER.isLoggable(Level.FINE)) {
-            LOGGER.log(Level.FINE, "for {0} have {1} of length {2} after upload of ‘{3}’", new Object[] {getId(), this.fileName, unencrypted().length, name});
+            LOGGER.log(Level.FINE, "for {0} have {1} of length {2} after upload of ‘{3}’",
+                    new Object[]{getId(), this.fileName, this.secretBytes.getPlainData().length, name});
         }
     }
 
-    @Override public String getFileName() {
-        return fileName;
+    /**
+     * Constructor for Stapler form binding.
+     *
+     * @param scope       the scope of the credentials.
+     * @param id          the id of the credentials.
+     * @param description the description of the credentials.
+     * @param file        the uploaded file.
+     * @param fileName    the name of the file.
+     * @param secretBytes the content of the file.
+     * @throws IOException when things go wrong.
+     */
+    @DataBoundConstructor
+    public FileCredentialsImpl(@CheckForNull CredentialsScope scope, @CheckForNull String id,
+                               @CheckForNull String description, @Nonnull FileItem file, @CheckForNull String fileName,
+                               @CheckForNull SecretBytes secretBytes) throws IOException {
+        super(scope, id, description);
+        String name = file.getName();
+        if (name.length() > 0) {
+            this.fileName = name.replaceFirst("^.+[/\\\\]", "");
+            this.secretBytes = SecretBytes.fromBytes(file.get());
+        } else {
+            this.fileName = fileName;
+            this.secretBytes = secretBytes;
```

Comment:
Recommended to do some assertions here: in the first clause, `secretBytes` should be null; in the second, should be non-null. (Right?)